File: src/algorithm/template/detection/base/file_runner.py
from src.algorithm.template.detection.base.common_pass import FilePassAggregates
from src.algorithm.template.detection.base.detection_pass import FieldTypeDetectionPass
from src.algorithm.template.detection.base.interpretation_pass import (
    FieldTypeInterpretationPass,
)
from src.algorithm.template.template_layout import Field


class FileLevelRunner:
    def __init__(
        self,
        tuning_config,
        detection_pass: FilePassAggregates,
        interpretation_pass: FilePassAggregates,
    ):
        self.tuning_config = tuning_config
        self.detection_pass = detection_pass
        self.interpretation_pass = interpretation_pass

    # Common wrappers
    def initialize_directory_level_aggregates(self, initial_directory_path):
        self.detection_pass.initialize_directory_level_aggregates(
            initial_directory_path
        )
        self.interpretation_pass.initialize_directory_level_aggregates(
            initial_directory_path
        )

    # TODO: see if this is really needed?
    # def get_aggregates_data(self):
    #     return {
    #         "detection": {
    #             "directory_level_aggregates": self.detection_pass.get_directory_level_aggregates(),
    #             # ...
    #         },
    #         "interpretation": {
    #             "directory_level_aggregates": self.interpretation_pass.get_directory_level_aggregates(),
    #             # ...
    #         },
    #         # More passes in future?
    #     }

    # Detection: Field Level
    def get_field_level_detection_aggregates(self):
        return self.detection_pass.get_field_level_aggregates()

    # Detection: File Level
    def initialize_file_level_detection_aggregates(self, file_path):
        return self.detection_pass.initialize_file_level_aggregates(file_path)

    def get_file_level_detection_aggregates(self):
        return self.detection_pass.get_file_level_aggregates()

    def update_detection_aggregates_on_processed_file(self, file_path):
        return self.detection_pass.update_aggregates_on_processed_file(file_path)

    # Detection: Directory Level
    def get_directory_level_detection_aggregates(self):
        return self.detection_pass.get_directory_level_aggregates()

    # Interpretation: Field Level
    def get_field_level_interpretation_aggregates(self):
        return self.interpretation_pass.get_field_level_aggregates()

    # Interpretation: File Level
    def initialize_file_level_interpretation_aggregates(
        self,
        file_path,
        field_detection_type_wise_detection_aggregates,
        field_label_wise_detection_aggregates,
    ):
        return self.interpretation_pass.initialize_file_level_aggregates(
            file_path,
            field_detection_type_wise_detection_aggregates,
            field_label_wise_detection_aggregates,
        )

    def get_file_level_interpretation_aggregates(self):
        return self.interpretation_pass.get_file_level_aggregates()

    def update_interpretation_aggregates_on_processed_file(self, file_path):
        return self.interpretation_pass.update_aggregates_on_processed_file(file_path)

    # Interpretation: Directory Level
    def get_directory_level_interpretation_aggregates(self):
        return self.interpretation_pass.get_directory_level_aggregates()


class FieldTypeFileLevelRunner(FileLevelRunner):
    """
    FieldTypeFileLevelRunner contains the external contract to be used by TemplateFileRunner for each of the field_detection_types
    It is static per template instance. Instantiated once per field type from the template.json

    """

    def __init__(
        self,
        tuning_config,
        field_detection_type,
        detection_pass: FieldTypeDetectionPass,
        interpretation_pass: FieldTypeInterpretationPass,
    ):
        self.field_detection_type = field_detection_type
        self.detection_pass: FieldTypeDetectionPass = None
        self.interpretation_pass: FieldTypeInterpretationPass = None
        super().__init__(tuning_config, detection_pass, interpretation_pass)

    def run_field_level_detection(self, field, gray_image, colored_image):
        self.detection_pass.initialize_field_level_aggregates(field)

        field_detection = self.detection_pass.get_field_detection(
            field, gray_image, colored_image
        )

        self.detection_pass.update_aggregates_on_processed_field_detection(
            field, field_detection
        )

        return field_detection

    def run_field_level_interpretation(self, field: Field):
        self.interpretation_pass.initialize_field_level_aggregates(field)
        file_level_detection_aggregates = (
            self.detection_pass.get_file_level_aggregates()
        )
        file_level_interpretation_aggregates = (
            self.interpretation_pass.get_file_level_aggregates()
        )
        # TODO: directly call .run_field_interpretation()
        field_interpretation = self.interpretation_pass.get_field_interpretation(
            field,
            file_level_detection_aggregates,
            file_level_interpretation_aggregates,
        )

        self.interpretation_pass.update_aggregates_on_processed_field_interpretation(
            field, field_interpretation
        )

        return field_interpretation

File: omrchecker-js/packages/core/src/processors/detection/base/fileRunner.ts
/**
 * File runner base classes.
 *
 * TypeScript port of src/processors/detection/base/file_runner.py
 * Coordinates detection and interpretation passes at file level.
 */

import cv from '../../../utils/opencv';
import { FilePassAggregates, type DirectoryLevelAggregates, type FileLevelAggregates, type FieldLevelAggregates } from './commonPass';
import { FieldTypeDetectionPass } from './detectionPass';
import { FieldTypeInterpretationPass } from './interpretationPass';
import type { Field } from '../../layout/field/base';
import type { FieldDetection } from './detection';
import type { FieldInterpretation } from './interpretation';
import type { TuningConfig } from './commonPass';

/**
 * Generic file-level runner coordinating detection and interpretation passes.
 *
 * Manages aggregates at different levels (file, directory) for both
 * detection and interpretation processes.
 */
export class FileLevelRunner<
  DetectionPassT extends FilePassAggregates,
  InterpretationPassT extends FilePassAggregates
> {
  protected tuningConfig: TuningConfig;
  protected detectionPass: DetectionPassT;
  protected interpretationPass: InterpretationPassT;

  constructor(
    tuningConfig: TuningConfig,
    detectionPass: DetectionPassT,
    interpretationPass: InterpretationPassT
  ) {
    this.tuningConfig = tuningConfig;
    this.detectionPass = detectionPass;
    this.interpretationPass = interpretationPass;
  }

  /**
   * Initialize directory-level aggregates for both detection and interpretation.
   *
   * @param initialDirectoryPath - Path to the initial directory
   */
  initializeDirectoryLevelAggregates(initialDirectoryPath: string): void {
    this.detectionPass.initializeDirectoryLevelAggregates(initialDirectoryPath);
    this.interpretationPass.initializeDirectoryLevelAggregates(initialDirectoryPath);
  }

  // Detection: Field Level
  /**
   * Get field-level detection aggregates.
   *
   * @returns Field-level detection aggregates
   */
  getFieldLevelDetectionAggregates(): FieldLevelAggregates | undefined {
    return this.detectionPass.getFieldLevelAggregates();
  }

  // Detection: File Level
  /**
   * Initialize file-level detection aggregates.
   *
   * @param filePath - Path to the file being processed
   */
  initializeFileLevelDetectionAggregates(filePath: string): void {
    this.detectionPass.initializeFileLevelAggregates(filePath);
  }

  /**
   * Get file-level detection aggregates.
   *
   * @returns File-level detection aggregates
   */
  getFileLevelDetectionAggregates(): FileLevelAggregates | undefined {
    return this.detectionPass.getFileLevelAggregates();
  }

  /**
   * Update detection aggregates when a file has been processed.
   *
   * @param filePath - Path to the processed file
   */
  updateDetectionAggregatesOnProcessedFile(filePath: string): void {
    this.detectionPass.updateAggregatesOnProcessedFile(filePath);
  }

  // Detection: Directory Level
  /**
   * Get directory-level detection aggregates.
   *
   * @returns Directory-level detection aggregates
   */
  getDirectoryLevelDetectionAggregates(): DirectoryLevelAggregates | undefined {
    return this.detectionPass.getDirectoryLevelAggregates();
  }

  // Interpretation: Field Level
  /**
   * Get field-level interpretation aggregates.
   *
   * @returns Field-level interpretation aggregates
   */
  getFieldLevelInterpretationAggregates(): FieldLevelAggregates | undefined {
    return this.interpretationPass.getFieldLevelAggregates();
  }

  // Interpretation: File Level
  /**
   * Initialize file-level interpretation aggregates.
   *
   * @param filePath - Path to the file being processed
   */
  initializeFileLevelInterpretationAggregates(filePath: string): void {
    this.interpretationPass.initializeFileLevelAggregates(filePath);
  }

  /**
   * Get file-level interpretation aggregates.
   *
   * @returns File-level interpretation aggregates
   */
  getFileLevelInterpretationAggregates(): FileLevelAggregates | undefined {
    return this.interpretationPass.getFileLevelAggregates();
  }

  /**
   * Update interpretation aggregates when a file has been processed.
   *
   * @param filePath - Path to the processed file
   */
  updateInterpretationAggregatesOnProcessedFile(filePath: string): void {
    this.interpretationPass.updateAggregatesOnProcessedFile(filePath);
  }

  // Interpretation: Directory Level
  /**
   * Get directory-level interpretation aggregates.
   *
   * @returns Directory-level interpretation aggregates
   */
  getDirectoryLevelInterpretationAggregates(): DirectoryLevelAggregates | undefined {
    return this.interpretationPass.getDirectoryLevelAggregates();
  }
}

/**
 * Field type file-level runner.
 *
 * Specializes FileLevelRunner for specific field types.
 * Handles detection and interpretation for specific fields.
 *
 * It contains the external contract to be used by TemplateFileRunner
 * for each of the field_detection_types.
 * It is static per template instance. Instantiated once per field type
 * from the template.json
 */
export class FieldTypeFileLevelRunner extends FileLevelRunner<
  FieldTypeDetectionPass,
  FieldTypeInterpretationPass
> {
  public fieldDetectionType: string;

  constructor(
    tuningConfig: TuningConfig,
    fieldDetectionType: string,
    detectionPass: FieldTypeDetectionPass,
    interpretationPass: FieldTypeInterpretationPass
  ) {
    super(tuningConfig, detectionPass, interpretationPass);
    this.fieldDetectionType = fieldDetectionType;
  }

  /**
   * Run field-level detection.
   *
   * @param field - Field to detect
   * @param grayImage - Grayscale image
   * @param coloredImage - Colored image (optional)
   * @returns FieldDetection result
   */
  runFieldLevelDetection(
    field: Field,
    grayImage: cv.Mat,
    coloredImage?: cv.Mat
  ): FieldDetection {
    // initializeFieldLevelAggregates is now called automatically by runFieldLevelDetection
    return this.detectionPass.runFieldLevelDetection(
      field,
      grayImage,
      coloredImage
    );
  }

  /**
   * Run field-level interpretation.
   *
   * @param field - Field to interpret
   * @returns FieldInterpretation result
   */
  runFieldLevelInterpretation(field: Field): FieldInterpretation {
    // initializeFieldLevelAggregates is now called automatically by runFieldLevelInterpretation
    const fileLevelDetectionAggregates =
      this.detectionPass.getFileLevelAggregates();
    return this.interpretationPass.runFieldLevelInterpretation(
      field,
      fileLevelDetectionAggregates
    );
  }
}

